src/librustc_hir/def.rs

```diff
@@ -451,13 +451,14 @@ impl<Id> Res<Id> {
         }
     }
 
-    pub fn matches_ns(&self, ns: Namespace) -> bool {
+    /// Returns `None` if this is `Res::Err`
+    pub fn ns(&self) -> Option<Namespace> {
         match self {
-            Res::Def(kind, ..) => kind.ns() == Some(ns),
-            Res::PrimTy(..) | Res::SelfTy(..) | Res::ToolMod => ns == Namespace::TypeNS,
-            Res::SelfCtor(..) | Res::Local(..) => ns == Namespace::ValueNS,
-            Res::NonMacroAttr(..) => ns == Namespace::MacroNS,
-            Res::Err => true,
+            Res::Def(kind, ..) => kind.ns(),
+            Res::PrimTy(..) | Res::SelfTy(..) | Res::ToolMod => Some(Namespace::TypeNS),
+            Res::SelfCtor(..) | Res::Local(..) => Some(Namespace::ValueNS),
+            Res::NonMacroAttr(..) => Some(Namespace::MacroNS),
+            Res::Err => None,
```

Comment:
Could you add `ns`, but keep `matches_ns` (perhaps reimplemented in terms of `ns`)?
The current version mistreats `Res::Err`.

Reply:
`ns` doesn't really make sense in presence of `Res::Err` it's certainly not `None`, it's, well, "All".

Reply:
That makes sense, it certainly simplifies the logic in `librustc_ast` quite a bit. Rustdoc never calls `ns()` on a `Res::Err` so from that perspective I don't mind what it returns. I added back `matches_ns`.